File: app/src/main/java/com/example/purepawapp/ui/spa/AppointmentsFragment.java
package com.example.purepawapp.ui.spa;

import android.os.Bundle;
import android.view.View;

import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.Booking;
import com.example.purepawapp.data.repository.RepoCallback;
import com.example.purepawapp.databinding.FragmentAppointmentsBinding;
import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.notification.NotificationHelper;
import com.example.purepawapp.ui.admin.AdminStatusUi;
import com.example.purepawapp.ui.admin.StatusStyle;
import com.example.purepawapp.ui.common.BaseFragment;
import com.example.purepawapp.util.BookingStatus;
import com.example.purepawapp.util.CurrencyUtils;
import com.example.purepawapp.util.ViewUtils;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.firebase.firestore.ListenerRegistration;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AppointmentsFragment extends BaseFragment<FragmentAppointmentsBinding> {

    private final AppointmentAdapter adapter = new AppointmentAdapter(
            booking -> {
                Bundle args = new Bundle();
                args.putString("bookingId", booking.getId());
                NavHostFragment.findNavController(this).navigate(R.id.action_appointmentsFragment_to_appointmentRatingFragment, args);
            },
            this::confirmCancel
    );

    private final Map<String, String> lastKnownStatuses = new HashMap<>();
    private ListenerRegistration bookingsListener;
    private boolean firstLoad = true;

    public AppointmentsFragment() {
        super(FragmentAppointmentsBinding::inflate);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getBinding().btnBack.setOnClickListener(v -> NavHostFragment.findNavController(this).popBackStack());
        getBinding().rvHistory.setLayoutManager(new LinearLayoutManager(requireContext()));
        getBinding().rvHistory.setAdapter(adapter);

        listenToBookings();
    }

    @Override
    public void onDestroyView() {
        if (bookingsListener != null) bookingsListener.remove();
        bookingsListener = null;
        super.onDestroyView();
    }

    private void listenToBookings() {
        String userId = ServiceLocator.getSessionManager().getUserIdOnce();
        if (userId == null || userId.isBlank()) return;

        showLoading();
        bookingsListener = ServiceLocator.getSpaRepository().listenToBookings(userId, new RepoCallback<>() {
            @Override
            public void onSuccess(List<Booking> bookings) {
                if (getBinding() == null) return;
                notifyStatusChanges(bookings);
                adapter.submitList(bookings);
                getBinding().tvEmpty.setVisibility(bookings.isEmpty() ? View.VISIBLE : View.GONE);
                getBinding().rvHistory.setVisibility(bookings.isEmpty() ? View.GONE : View.VISIBLE);
                renderStats(bookings);
                if (firstLoad) {
                    hideLoading();
                    firstLoad = false;
                }
            }

            @Override
            public void onError(Exception error) {
                hideLoading();
                ViewUtils.toast(AppointmentsFragment.this, error.getMessage() != null ? error.getMessage() : "Không thể tải lịch sử đặt lịch");
            }
        });
    }

    private void notifyStatusChanges(List<Booking> bookings) {
        if (!firstLoad) {
            for (Booking booking : bookings) {
                String previous = lastKnownStatuses.get(booking.getId());
                if (previous != null && !previous.equals(booking.getStatus())) {
                    StatusStyle style = AdminStatusUi.bookingStatusStyle(booking.getStatus());
                    NotificationHelper.showNotification(requireContext(),
                            "Lịch hẹn " + booking.getServiceName() + " cập nhật",
                            "Trạng thái mới: " + style.getLabel(),
                            booking.getId().hashCode());
                }
            }
        }
        lastKnownStatuses.clear();
        for (Booking booking : bookings) lastKnownStatuses.put(booking.getId(), booking.getStatus());
    }

    private void confirmCancel(Booking booking) {
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Hủy lịch hẹn")
                .setMessage("Bạn có chắc muốn hủy lịch \"" + booking.getServiceName() + "\"?")
                .setNegativeButton("Không", null)
                .setPositiveButton("Hủy lịch", (dialog, which) -> cancelBooking(booking))
                .show();
    }

    private void cancelBooking(Booking booking) {
        showLoading();
        ServiceLocator.getSpaRepository().updateBookingStatus(booking.getId(), BookingStatus.CANCELLED, new RepoCallback<>() {
            @Override
            public void onSuccess(Void result) {
                hideLoading();
                ViewUtils.toast(AppointmentsFragment.this, "Đã hủy lịch hẹn");
            }

            @Override
            public void onError(Exception error) {
                hideLoading();
                ViewUtils.toast(AppointmentsFragment.this,
                        error.getMessage() != null ? error.getMessage() : "Không thể hủy lịch, vui lòng thử lại");
            }
        });
    }

    private void renderStats(List<Booking> bookings) {
        int count = bookings.size();
        double totalSpent = 0;
        int ratingSum = 0;
        int ratedCount = 0;
        for (Booking booking : bookings) {
            if (BookingStatus.COMPLETED.equals(booking.getStatus())) totalSpent += booking.getPrice();
            if (booking.getRating() > 0) {
                ratingSum += booking.getRating();
                ratedCount++;
            }
        }
        getBinding().tvStatCount.setText(String.valueOf(count));
        getBinding().tvStatSpent.setText(CurrencyUtils.toVndString(totalSpent));
        double avgRating = ratedCount == 0 ? 0.0 : (double) ratingSum / ratedCount;
        getBinding().tvStatRating.setText(String.format("%.1f⭐", avgRating));
    }
}

File: app/src/main/java/com/example/purepawapp/ui/spa/AppointmentsFragment.kt
package com.example.purepawapp.ui.spa

import android.content.res.ColorStateList
import android.os.Bundle
import android.view.View
import androidx.core.os.bundleOf
import androidx.navigation.fragment.findNavController
import com.example.purepawapp.R
import com.example.purepawapp.databinding.FragmentAppointmentsBinding
import com.example.purepawapp.databinding.ItemAppointmentHistoryBinding
import com.example.purepawapp.ui.common.BaseFragment

class AppointmentsFragment : BaseFragment<FragmentAppointmentsBinding>(FragmentAppointmentsBinding::inflate) {

    private data class HistoryEntry(
        val bookingId: String,
        val date: String,
        val code: String,
        val service: String,
        val pet: String,
        val price: String,
        val iconEmoji: String,
        val iconBgColor: Int,
        val rating: Int
    )

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btnBack.setOnClickListener {
            findNavController().popBackStack()
        }

        val entries = listOf(
            HistoryEntry(
                bookingId = "SP230045",
                date = "📅 10/06/2026 · 10:00–12:00",
                code = "SP230045",
                service = "Spa & Grooming",
                pet = "Lucky (Poodle) · Q.1",
                price = "250.000đ",
                iconEmoji = "✂️",
                iconBgColor = R.color.pp_chip_pink_bg,
                rating = 5
            ),
            HistoryEntry(
                bookingId = "MK230031",
                date = "📅 02/06/2026 · 14:00–15:00",
                code = "MK230031",
                service = "Khám sức khỏe định kỳ",
                pet = "Mochi (Mèo Anh) · Q.3",
                price = "150.000đ",
                iconEmoji = "🏥",
                iconBgColor = R.color.pp_chip_green_bg,
                rating = 4
            ),
            HistoryEntry(
                bookingId = "SP230018",
                date = "📅 18/05/2026 · 09:00–10:00",
                code = "SP230018",
                service = "Tắm & Sấy",
                pet = "Lucky (Poodle) · Q.1",
                price = "120.000đ",
                iconEmoji = "💧",
                iconBgColor = R.color.pp_chip_orange_bg,
                rating = 5
            ),
            HistoryEntry(
                bookingId = "MK230009",
                date = "📅 05/05/2026 · 11:00–11:30",
                code = "MK230009",
                service = "Tiêm phòng dại",
                pet = "Lucky (Poodle) · Q.BT",
                price = "80.000đ",
                iconEmoji = "💉",
                iconBgColor = R.color.pp_chip_blue_bg,
                rating = 5
            )
        )

        val rows = listOf(
            binding.itemHistory1,
            binding.itemHistory2,
            binding.itemHistory3,
            binding.itemHistory4
        )

        rows.zip(entries).forEach { (row, entry) -> bind(row, entry) }
    }

    private fun bind(row: ItemAppointmentHistoryBinding, entry: HistoryEntry) {
        row.tvDate.text = entry.date
        row.tvCode.text = entry.code
        row.tvService.text = entry.service
        row.tvPet.text = entry.pet
        row.tvPrice.text = entry.price
        row.tvIcon.text = entry.iconEmoji
        row.tvIcon.backgroundTintList = ColorStateList.valueOf(resources.getColor(entry.iconBgColor, null))
        row.tvStars.text = "⭐".repeat(entry.rating) + "☆".repeat(5 - entry.rating)
        row.root.setOnClickListener {
            findNavController().navigate(
                R.id.action_appointmentsFragment_to_appointmentRatingFragment,
                bundleOf("bookingId" to entry.bookingId)
            )
        }
    }
}
